### DISCUSSION TITLE
Conditional Group By

### DISCUSSION THREAD

**Discussion by MohMehrnia** [ravendb/ravendb#18260] [General]
Why this code not work?

```
from "VideoPlayUserInfos" 
where State='StartPlay'
group by UserName
select UserName, count(UserName) as  co
```

**Comment by ayende** [comment]
The issue is that you are trying to first filter the results then group them, but that would mean that we need to do the aggregation on the fly.

Instead, write the query like so:
```
from "VideoPlayUserInfos" 
group by UserName, State
where State='StartPlay'
select UserName, count() as  co
```

That allows RavenDB to do the appropriate aggregations ahead of time and give you the results quickly, regardless of the data size.

**Comment by MohMehrnia** [comment]
Thanks, I need the below scenario

group by state, sort with log date, and select first record  in each record

My Sample Document is 

```
{
    "UserID": "eaba13f6-3acd-4acc-be2f-0cacb219e908",
    "FileID": "52a840a4-0099-49ca-acf2-11a3e8c58f42",
    "LogDateTime": "2024-03-22T21:16:01.7892262",
    "CurrentTime": 5.09628,
    "UserName": "moh.mehrnia",
    "ContentDuration": "00:02:14.3319777",
    "State": "Pause",
    "FullName": "محمد مهرنیا",
    "@metadata": {
        "@collection": "VideoPlayUserInfos",
        "Raven-Clr-Type": "Sin.Course.DTO.JsonObjects.VideoPlayUserInfo, Sin.Course.DTO"
    }
}
```

**Reply by ayende** [reply]
Assuming this document structure, you can define a map/reduce index like so:

```
// map
from v in docs.VideoPlayUserInfos
// where v.State == "StartPlay" // if you want to filter at the source
select new { v.State, v.UserName, Count = 1, FirstSeen = v.LogDateTime }

// map
from r in results
group r by new { r.State, r.UserName } into g
select new {
   g.Key.State,
   g.Key.UserName,
   FirstSeen = g.OrderByDescending(x=>x.FirstSeen).First().FirstSeen,
   Count = g.Sum(x=>x.Count)
}
```

And then you can do that immediately

**Reply by MohMehrnia** [reply]
how do created in c# and another question (thanks for helping)

another question 
how do query by map/reduce index

**Reply by ayende** [reply]
See here:

https://ravendb.net/docs/article-page/6.0/csharp/indexes/creating-and-deploying

https://ravendb.net/docs/article-page/6.0/csharp/indexes/map-reduce-indexes#creating-map-reduce-indexes

https://ravendb.net/docs/article-page/6.0/csharp/indexes/querying/query-index#session.query

**Comment by ayende** [comment]
Probably this:

https://ayende.com/blog/197637-A/production-postmortem-your-math-is-wrong-recursion-doesnt-work-this-way

On Mon, Apr 8, 2024 at 2:56 AM Mohammad Mehrnia ***@***.***>
wrote:

> Why some record not in results!!!!!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ravendb/ravendb/discussions/18260#discussioncomment-9039319>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4RM4R3VUSWVSR2QL4HTTY4HMJ5AVCNFSM6AAAAABE24R6USVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TAMZZGMYTS>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
<https://ravendb.net/> *OREN EINI*
*CEO & Founder*
***@***.***
+972-52-548-6969
www.ravendb.net <https://ravendb.net/>

<https://www.linkedin.com/company/ravendb/>  <https://twitter.com/ravendb>
<https://github.com/ravendb>
<https://ravendb.net/emailsignature/displayeventpage>
